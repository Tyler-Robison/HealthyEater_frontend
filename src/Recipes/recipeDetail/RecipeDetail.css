.RecipeDetail{
    background-image: url('../../Images/food_background.jpg');
    background-repeat: no-repeat;
    background-size: cover;
    background-attachment: fixed;
    display: flex;
    flex-direction: column;
    align-items: center;
}



.RecipeDetail-summary {
    background-color: white;
    width: 40vw;
    border-radius: 1rem;
    box-shadow: 2px 2px 8px rgba(0, 0, 0, 1);
}

.RecipeDetail-summary p {
    margin: 0.4rem 0;
    font-size: 1.4rem;
}

.step-div {
    background-color: rgb(217, 240, 248);
    border: 2px solid black;
    margin-bottom: 1rem;
    box-shadow: 10px 5px 5px rgb(87, 87, 87);
}

@media (max-width: 800px) {
    .RecipeDetail-summary {
        font-size: 0.8rem;
    }

    .RecipeDetail-summary p {
        margin: 0.4rem 0;
        font-size: 1rem;
    }

    .RecipeDetail-header {
        font-size: 20px;
    }

    .step-div p{
       font-size: 0.8rem;
    }
}

@media (max-width: 468px) {
    .RecipeDetail-summary {
        font-size: 0.8rem;
    }

    .RecipeDetail-summary p {
        margin: 0.4rem 0;
        font-size: 0.8rem;
    }

    .RecipeDetail-header {
        font-size: 14px;
    }


}



.steps-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-top: 1rem;
}

.go-back-btn {
    background-color: red;
    color: white;
    font-weight: 700;
    padding: 0.5rem 1rem;
    border-radius: 10%;
    margin-top: 1rem;
}

.save-recipe-btn {
    margin-top: 0.8rem;
}

.header-backer {
    background-color: white;
    border: 1px solid black;
    box-shadow: 2px 2px 8px rgba(0, 0, 0, 1);
}

/* .steps-container .step-number { */
    /* align-self: flex-end; */
    /* color: aqua; */
/* } */


.nutrition-div {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.StepList {
    max-width: 600px;
}