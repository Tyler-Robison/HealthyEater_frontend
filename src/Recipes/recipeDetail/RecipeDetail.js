import React, { useEffect, useState, useContext } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import SpoonacularAPI from '../../APIs/spoonAPI';
import RecipeAPI from '../../APIs/recipeAPI';
import NutrientTable from './NutrientTable';
import StepsList from './StepsList';
import GlobalContext from '../../context/GlobalContext';
import Spinner from '../../Spinner';
import './RecipeDetail.css'

const RecipeDetail = () => {
    const { currentUser, setCurrentUser, token } = useContext(GlobalContext)
    const { recipeID } = useParams()
    const navigate = useNavigate();
    const [recipeDetail, setRecipeDetail] = useState(null)
    const [nutritionDetail, setNutritionDetail] = useState(null)

    // Needs to direct user to page they came from
    // can be recipeList, Mealplan or savedRecipes
    const goBack = () => navigate(-1)

    useEffect(() => {
        const getRecipeDetail = async () => {
            const res = await SpoonacularAPI.getRecipeDetail(recipeID, token, currentUser.id);
            setNutritionDetail(res.nutrition)
            setRecipeDetail(res.recipe)
        }
        getRecipeDetail();
    }, [])

    // only care recipe detail, nutri detail can be obtained based on recipe_id
    const saveRecipe = async () => {
        const recipeRes = await RecipeAPI.saveRecipe(recipeDetail, currentUser.id, token)

        currentUser.recipes.push(recipeRes.savedRecipe)

        setCurrentUser({ ...currentUser, currentUser })
    }

    if (!recipeDetail) return <Spinner />

    return (
        <div className='RecipeDetail'>
            <div className='row mt-3'>
                <div className='RecipeDetail-summary col-12'>
                    <h1 className='RecipeDetail-header'>{recipeDetail.title}</h1>

                    <p>Preparation Time: {recipeDetail.readyInMinutes} Minutes</p>
                    <p>Serves: {recipeDetail.servings}</p>
                    <p>Weight Watcher's Smart Points: {recipeDetail.weightWatcherSmartPoints}</p>
                </div>
            </div>
            <button className='general-btn save-recipe-btn' onClick={saveRecipe}>Save Recipe</button>
            <button className='general-btn-red' onClick={goBack}>Back</button>
            <div className='steps-container'>
                <StepsList recipeDetail={recipeDetail} />
            </div>

            <div className='nutrition-div'>
                <div className='header-backer'><h2>Bad Stuff!</h2></div>
                <NutrientTable nutritionDetail={nutritionDetail.bad} />

                <div className='header-backer'><h2>Good Stuff!</h2></div>
                <NutrientTable nutritionDetail={nutritionDetail.good} />
            </div>

            <button className='general-btn-red' onClick={goBack}>Back</button>
        </div>
    )

}

export default RecipeDetail;