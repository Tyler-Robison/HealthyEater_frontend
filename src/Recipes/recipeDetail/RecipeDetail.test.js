import React from 'react';
import { render, screen } from '@testing-library/react';
import RecipeDetail from './RecipeDetail';
import { MemoryRouter } from "react-router-dom";
import ContextProvider from '../../testContext';
import { waitFor, cleanup, waitForElementToBeRemoved } from '@testing-library/react';
import mockedAxios from 'axios';


const goodNutrients = [
    { title: 'Protein', amount: '68g', indented: false, percentOfDailyNeeds: 137.37 },
    { title: 'Selenium', amount: '105µg', indented: false, percentOfDailyNeeds: 150.87 },
    { title: 'Vitamin B3', amount: '28mg', indented: false, percentOfDailyNeeds: 143.89 },
    { title: 'Vitamin B6', amount: '1mg', indented: false, percentOfDailyNeeds: 97.74 },
    { title: 'Phosphorus', amount: '758mg', indented: false, percentOfDailyNeeds: 75.9 },
    { title: 'Vitamin B1', amount: '0.85mg', indented: false, percentOfDailyNeeds: 56.5 },
    { title: 'Vitamin B2', amount: '0.77mg', indented: false, percentOfDailyNeeds: 45.51 },
    { title: 'Manganese', amount: '0.83mg', indented: false, percentOfDailyNeeds: 41.55 },
    { title: 'Vitamin B5', amount: '4mg', indented: false, percentOfDailyNeeds: 40.63 },
    { title: 'Vitamin K', amount: '42µg', indented: false, percentOfDailyNeeds: 40.47 },
    { title: 'Iron', amount: '6mg', indented: false, percentOfDailyNeeds: 37.35 },
    { title: 'Potassium', amount: '1143mg', indented: false, percentOfDailyNeeds: 32.67 },
    { title: 'Folate', amount: '121µg', indented: false, percentOfDailyNeeds: 30.38 },
    { title: 'Calcium', amount: '277mg', indented: false, percentOfDailyNeeds: 27.78 },
    { title: 'Magnesium', amount: '107mg', indented: false, percentOfDailyNeeds: 26.77 },
    { title: 'Zinc', amount: '3mg', indented: false, percentOfDailyNeeds: 23.98 },
    { title: 'Vitamin B12', amount: '1µg', indented: false, percentOfDailyNeeds: 17.71 },
    { title: 'Fiber', amount: '3g', indented: false, percentOfDailyNeeds: 15.3 },
    { title: 'Copper', amount: '0.26mg', indented: false, percentOfDailyNeeds: 13.16 },
    { title: 'Vitamin A', amount: '588IU', indented: false, percentOfDailyNeeds: 11.77 },
    { title: 'Vitamin C', amount: '6mg', indented: false, percentOfDailyNeeds: 8.31 },
    { title: 'Vitamin E', amount: '1mg', indented: false, percentOfDailyNeeds: 7.09 },
    { title: 'Vitamin D', amount: '0.96µg', indented: false, percentOfDailyNeeds: 6.43 }
]

const badNutrients = [
    { title: 'Calories', amount: '660k', indented: false, percentOfDailyNeeds: 33.03 },
    { title: 'Fat', amount: '20g', indented: false, percentOfDailyNeeds: 31.41 },
    { title: 'Saturated Fat', amount: '7g', indented: true, percentOfDailyNeeds: 48.21 },
    { title: 'Carbohydrates', amount: '46g', indented: false, percentOfDailyNeeds: 15.51 },
    { title: 'Sugar', amount: '2g', indented: true, percentOfDailyNeeds: 2.39 },
    { title: 'Cholesterol', amount: '261mg', indented: false, percentOfDailyNeeds: 87.23 },
    { title: 'Sodium', amount: '1134mg', indented: false, percentOfDailyNeeds: 49.32 }
]

const mockInstructions = {
    steps: [
        {
            number: 1,
            step: 'Do the first step',
            equipment: [{ name: 'mock equipment' }, { name: 'other equipment' }],
            ingredients: [{ name: 'mock ingredient' }, { name: 'other ingredient' }]
        },
        {
            number: 2,
            step: 'Do the second step',
            equipment: [],
            ingredients: []
        },
        {
            number: 3,
            step: 'Do the third step',
            equipment: [{ name: 'mock equipment' }],
            ingredients: [{ name: 'mock ingredient' }]
        }
    ]
}

const data = {
    data: {
        nutrition: {
            good: goodNutrients,
            bad: badNutrients
        },
        recipe: {
            title: 'mocked recipe title',
            readyInMinutes: 45,
            weightWatcherSmartPoints: 4,
            servings: 5,
            analyzedInstructions: [mockInstructions]
        }
    }
};

// getting correct value back from mockAxios, but not waiting
// if (!recipeDetail) return <p>Loading...</p>

afterEach(cleanup);

it("renders without crashing", function () {
    mockedAxios.get.mockResolvedValueOnce(data);

        <MemoryRouter>
            <ContextProvider >
                <RecipeDetail />
            </ContextProvider>
        </MemoryRouter>
});


it('Matches snapshot', async () => {
    mockedAxios.get.mockResolvedValueOnce(data);

    const { asFragment } = render(
        <MemoryRouter>
            <ContextProvider >
                <RecipeDetail />
            </ContextProvider>
        </MemoryRouter>
    );

    // will get screenshot of the loading text unless we wait
    await waitForElementToBeRemoved(screen.getByText('Loading...'));

    expect(asFragment()).toMatchSnapshot();
})

// test('recipeDetail', async () => {

//     mockedAxios.get.mockResolvedValueOnce(data);

//     const detailComponent = render(
//         <MemoryRouter>
//             <ContextProvider >
//                 <RecipeDetail />
//             </ContextProvider>
//         </MemoryRouter>
//     );

//     await waitFor(() => {

//         expect(mockedAxios.get.mockResolvedValueOnce(data)).toHaveBeenCalledTimes(1)
//         expect(detailComponent.getByText('mocked recipe title'));
//     });
// });

