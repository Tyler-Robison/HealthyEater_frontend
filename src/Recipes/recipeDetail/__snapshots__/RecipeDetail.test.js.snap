// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="RecipeDetail"
  >
    <div
      class="row mt-3"
    >
      <div
        class="RecipeDetail-summary col-12"
      >
        <h1
          class="RecipeDetail-header"
        >
          mocked recipe title
        </h1>
        <p>
          Preparation Time: 45 Minutes
        </p>
        <p>
          Serves: 5
        </p>
        <p>
          Weight Watcher's Smart Points: 4
        </p>
      </div>
    </div>
    <button
      class="general-btn save-recipe-btn"
    >
      Save Recipe
    </button>
    <button
      class="general-btn-red"
    >
      Back
    </button>
    <div
      class="steps-container"
    >
      <div
        class="d-flex flex-column align-items-center row StepList"
      >
        <div
          class="step-div col-10"
        >
          <p
            class="step-number"
          >
            <b>
              1
            </b>
          </p>
          <p>
            Do the first step
          </p>
          <p>
            <b>
              Equipment Used:
            </b>
             mock equipment, other equipment 
          </p>
          <p>
            <b>
              Ingredients Used:
            </b>
             mock ingredient, other ingredient 
          </p>
        </div>
        <div
          class="step-div col-10"
        >
          <p
            class="step-number"
          >
            <b>
              2
            </b>
          </p>
          <p>
            Do the second step
          </p>
          <p>
            <b>
              Equipment Used:
            </b>
             None 
          </p>
          <p>
            <b>
              Ingredients Used:
            </b>
             None 
          </p>
        </div>
        <div
          class="step-div col-10"
        >
          <p
            class="step-number"
          >
            <b>
              3
            </b>
          </p>
          <p>
            Do the third step
          </p>
          <p>
            <b>
              Equipment Used:
            </b>
             mock equipment 
          </p>
          <p>
            <b>
              Ingredients Used:
            </b>
             mock ingredient 
          </p>
        </div>
      </div>
    </div>
    <div
      class="nutrition-div"
    >
      <div
        class="header-backer"
      >
        <h2>
          Bad Stuff!
        </h2>
      </div>
      <table
        class="nutrition-table"
      >
        <thead>
          <tr>
            <th>
              Nutrient
            </th>
            <th>
              Amount
            </th>
            <th>
              Daily Need (%)
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Calories
            </td>
            <td>
              660k
            </td>
            <td>
              33.03
            </td>
          </tr>
          <tr>
            <td>
              Fat
            </td>
            <td>
              20g
            </td>
            <td>
              31.41
            </td>
          </tr>
          <tr>
            <td>
              Saturated Fat
            </td>
            <td>
              7g
            </td>
            <td>
              48.21
            </td>
          </tr>
          <tr>
            <td>
              Carbohydrates
            </td>
            <td>
              46g
            </td>
            <td>
              15.51
            </td>
          </tr>
          <tr>
            <td>
              Sugar
            </td>
            <td>
              2g
            </td>
            <td>
              2.39
            </td>
          </tr>
          <tr>
            <td>
              Cholesterol
            </td>
            <td>
              261mg
            </td>
            <td>
              87.23
            </td>
          </tr>
          <tr>
            <td>
              Sodium
            </td>
            <td>
              1134mg
            </td>
            <td>
              49.32
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="header-backer"
      >
        <h2>
          Good Stuff!
        </h2>
      </div>
      <table
        class="nutrition-table"
      >
        <thead>
          <tr>
            <th>
              Nutrient
            </th>
            <th>
              Amount
            </th>
            <th>
              Daily Need (%)
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Protein
            </td>
            <td>
              68g
            </td>
            <td>
              137.37
            </td>
          </tr>
          <tr>
            <td>
              Selenium
            </td>
            <td>
              105µg
            </td>
            <td>
              150.87
            </td>
          </tr>
          <tr>
            <td>
              Vitamin B3
            </td>
            <td>
              28mg
            </td>
            <td>
              143.89
            </td>
          </tr>
          <tr>
            <td>
              Vitamin B6
            </td>
            <td>
              1mg
            </td>
            <td>
              97.74
            </td>
          </tr>
          <tr>
            <td>
              Phosphorus
            </td>
            <td>
              758mg
            </td>
            <td>
              75.9
            </td>
          </tr>
          <tr>
            <td>
              Vitamin B1
            </td>
            <td>
              0.85mg
            </td>
            <td>
              56.5
            </td>
          </tr>
          <tr>
            <td>
              Vitamin B2
            </td>
            <td>
              0.77mg
            </td>
            <td>
              45.51
            </td>
          </tr>
          <tr>
            <td>
              Manganese
            </td>
            <td>
              0.83mg
            </td>
            <td>
              41.55
            </td>
          </tr>
          <tr>
            <td>
              Vitamin B5
            </td>
            <td>
              4mg
            </td>
            <td>
              40.63
            </td>
          </tr>
          <tr>
            <td>
              Vitamin K
            </td>
            <td>
              42µg
            </td>
            <td>
              40.47
            </td>
          </tr>
          <tr>
            <td>
              Iron
            </td>
            <td>
              6mg
            </td>
            <td>
              37.35
            </td>
          </tr>
          <tr>
            <td>
              Potassium
            </td>
            <td>
              1143mg
            </td>
            <td>
              32.67
            </td>
          </tr>
          <tr>
            <td>
              Folate
            </td>
            <td>
              121µg
            </td>
            <td>
              30.38
            </td>
          </tr>
          <tr>
            <td>
              Calcium
            </td>
            <td>
              277mg
            </td>
            <td>
              27.78
            </td>
          </tr>
          <tr>
            <td>
              Magnesium
            </td>
            <td>
              107mg
            </td>
            <td>
              26.77
            </td>
          </tr>
          <tr>
            <td>
              Zinc
            </td>
            <td>
              3mg
            </td>
            <td>
              23.98
            </td>
          </tr>
          <tr>
            <td>
              Vitamin B12
            </td>
            <td>
              1µg
            </td>
            <td>
              17.71
            </td>
          </tr>
          <tr>
            <td>
              Fiber
            </td>
            <td>
              3g
            </td>
            <td>
              15.3
            </td>
          </tr>
          <tr>
            <td>
              Copper
            </td>
            <td>
              0.26mg
            </td>
            <td>
              13.16
            </td>
          </tr>
          <tr>
            <td>
              Vitamin A
            </td>
            <td>
              588IU
            </td>
            <td>
              11.77
            </td>
          </tr>
          <tr>
            <td>
              Vitamin C
            </td>
            <td>
              6mg
            </td>
            <td>
              8.31
            </td>
          </tr>
          <tr>
            <td>
              Vitamin E
            </td>
            <td>
              1mg
            </td>
            <td>
              7.09
            </td>
          </tr>
          <tr>
            <td>
              Vitamin D
            </td>
            <td>
              0.96µg
            </td>
            <td>
              6.43
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <button
      class="general-btn-red"
    >
      Back
    </button>
  </div>
</DocumentFragment>
`;
